### PR TITLE
CRLP: Fix issue with getting correct target currency in batch jobs with records in multiple currencies

### LINKED ISSUE #3631: Customizable Rollups: Not Getting Correct Currency and MultiCurrency Conversions May Be Incorrect. Batch Job Does Not Correctly Recalculate Values (But Recalculate Rollups Button does)
_Note: I tested this with both standard MultiCurrency and Advanced MultiCurrency and the behavior is the same._

If the Currency of an Account changes after Customizable Rollups have run, the rollup fields have incorrect values. For instance, if the Org's Corporate Currency is GBP, a Contact's Currency is USD, and their Household Account is GBP (this happens by default when the Contact is created with a Currency that is not the default) and then you create a Closed Won Opportunity with USD as the Currency, everything calculates properly. When the Household Account's Currency is then changed to USD, the values convert to totally incorrect values. Running the Rollup Donations Batch does not recalculate the values properly. Clicking the Recalculate Rollups button on that single Household Account, however, does convert it to the correct values.

There is similar behavior on legacy rollups. When switching the Account's Currency to USD after rollups have run, the values are incorrectly calculated. However, the rollup donations batch DOES switch them to the correct value (whereas it does not with customizable rollups).

Contacts seem to recalculate properly with the batch. (I ran a test with both Contact/Account as GBP and Opportunity as USD. Rollups ran and then I switched the Currency of both the Contact and Account to USD. The rollups had incorrect values. When I ran the batch, the Contact rollups were correct; the Account's were not.)

**Steps to Repeat**

*Initial Setup*

1. Create scratch org and run install_prod flow so that you can test with the managed package. [Unmanaged may be fine, but I was testing the BDI/CRLPS issue with managed]
2. Enable Customizable Rollups in NPSP Settings → Donations → Customizable Rollups.
3. Go to Setup —> Object Manager → Contact → Page Layouts → Page Layout Assignment —> Edit Assignments. Select all of the Profiles and choose the Contact Layout from our installed package in the drop down. Click Save.

*Activate Multiple Currencies*

1. Go to Setup→ Company Information → Edit
2. Default Locale = English (United Kingdom)
3. Currency Locale: English (United Kingdom)  - GBP
4. Activate multiple currencies and hit Save.
5. Click Currency Setup button in Setup → Company Information.
6. Add new Active Currency
    1. Currency Type = USD
    2. Conversion Rate = 1.29
    3. Decimal Places = 2
    4. Save
7. If you want to test with Advanced MultiCurrency (not necessary), Click Enable Advanced Currency Management button.

*Create the Data*

The Controls:

1. Contact is GBP; Household is GBP (since default org currency is GBP).
    1. Create a Contact (Judi Sohn) and make sure to set the Currency on the Contact to USD. The Account's Household will stay as the default (GBP). 
    2. Create an Opportunity from the Household  (Amount = 1000; Close Date = today; Stage = Closed Won; Currency = USD).
    3. The rollups will calculate properly. Total Gifts on the Contact show GBP 775.19 and Total Gifts on the Account show GBP 775.19. 
2. Contact is USD; Household is GBP (this happens by default when you create a Contact outside of the default Currency).
    1. Create a Contact (Cori O'Brien) and make sure to set the Currency on the Contact to USD. The Account's Household will stay as the default (GBP). 
    2. Create an Opportunity from the Household  (Amount = 1000; Close Date = today; Stage = Closed Won; Currency = USD).
    3. The rollups will calculate properly. Total Gifts on the Contact show USD 1,000 (GBP 775.19) and Total Gifts on the Account show GBP 775.192. 

The Bug:

1. Contact is USD; Household is GBP, but then you switch it to USD after rollups have already run.
    1. Create a Contact (Jess Lopez) and make sure to set the Currency on the Contact to USD. The Account's Household will stay as the default (GBP). 
    2. Create an Opportunity from the Household  (Amount = 1000; Close Date = today; Stage = Closed Won; Currency = USD).
    3. The rollups will calculate properly. Total Gifts on the Contact show USD 1,000 (GBP 775.19) and Total Gifts on the Account show GBP 775.192. 
    4. Now switch the Household Account's Currency to USD.
    5. The rollups switch to incorrect values on the Household Account [i.e. Total Gifts = USD 775.19 (GBP 600.92)]
    6. Run Rollup Donations Batch by going to NPSP Settings → Bulk Data Processes → Rollup. Donations Batch. The incorrect values stay the same on the Household  [i.e. Total Gifts = USD 775.19 (GBP 600.92)]
    7. Clear out the Total Gifts value on the Household Account and run the Rollup Donations batch again. It still incorrectly calculates the Total Gifts as USD 775.19 (GBP 600.92). I even tried clearing every rollup and it still calculated improperly.
    8. Now click the Recalculate Rollups button on the Household Record itself. It will convert the values properly  [i.e. Total Gifts = USD 1,000.00 (GBP 775.19)]

It's definitely something in the batch. I ran one additional test where I switched back to legacy rollups. The Contact was USD, the HH Account was GBP, and the Opportunity was USD. Rollups were correct. I then switched the HH Account to USD so the values were incorrect but then ran the Rollup Donations batch to get the correct values. Finally, I cleared out all of the rollup values, switched customizable rollups on, and then ran the Rollup Donations Batch. It incorrectly calculated the rollup values  [i.e. Total Gifts = USD 775.19 (GBP 600.92)].

### PR DESCRIPTION
**Pull request by force2b**
Be sure to set the Handler service currency code each time the handler is used rather than only the first time it's instantiated. In addition, apply that currency code to the rollup definition when resetting the rollup collection in the handler service before starting a new rollup calculation for a given record.

# Critical Changes

# Changes

# Issues Closed

Fixes #3631 

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by jesslopez-sf** [comment]
This looks great, @force2b. All of my tests worked as expected and the data that was previously incorrect was corrected once the batch with the updated code was run.

File: src/classes/CRLP_RollupProcessor_SVC.cls [hunk 1]
```diff
@@ -276,12 +276,13 @@ public class CRLP_RollupProcessor_SVC {
             Type handlerClassType = getHanderClassType(options);
             CRLP_Debug_UTIL.setCurrentRollupState('Instantiate Handler Class: ' + handlerClassType.getName() + ' for ' + parentId);
             handler = (CRLP_VRollupHandler) handlerClassType.newInstance();
-            if (UserInfo.isMultiCurrencyOrganization()) {
-                handler.setCurrencyCode((String)parent.get('CurrencyIsoCode'));
-            }
             handlersMapByType.put(options.rollupJobType, handler);
         }
```
[the remark this answers is not in the document]
Comment (force2b) [reply src/classes/CRLP_RollupProcessor_SVC.cls]: Good catch.

File: src/classes/CRLP_VRollupHandler.cls [hunk 2]
```diff
@@ -167,6 +167,7 @@ public virtual class CRLP_VRollupHandler implements CRLP_Rollup_SVC.IRollupHandl
     public void resetAllRollups() {
         for (CRLP_Rollup r : this.rollups) {
             r.resetValues();
+            r.setCurrencyCode(this.currCode);
```
[the remark this answers is not in the document]
Comment (force2b) [reply src/classes/CRLP_VRollupHandler.cls]: As a habit, I've always added `this.` to specify instance vars (pre-style doc) There's use of that pattern throughout all of the Customizable Rollups code so I'm not sure I want to pull it out of just this one method. If you're ok with that, I'd like to just leave this one as is.

**Comment by davisagli** [comment]
@force2b @dvenor This test is failing in the multicurrency beta org: https://mrbelvedereci.herokuapp.com/builds/37327/tests

**Comment by force2b** [comment]
I noticed that. It's because the multicurrency beta org only has a single currency in it, even though multicurrency is enabled. It's minor. The ideal fix is to have the org created with a second currency, but that may not be possible (you can't load through apex or as metadata). Worse case is that the unit test just checks to see if there is at least two currency codes before running the test.

**Comment by davisagli** [comment]
@force2b okay, I made a work item for me to look into configuring a second currency in that org. I do think it would be good to add that check before running the test.

**Comment by sam-knox** [comment]
@force2b - Is there anything the end user needs to do to enable this fix? Also does this only affect Account rollups and not Contact?

**Comment by force2b** [comment]
@sam-knox No. It's a fix in the processing logic. It could technically affect any rollup - Account, Contact, GAU, Recurring Donation, Hard Credit, Soft Credit, etc. - though realistically only Accounts & Contacts would have different currencies as target summary records.

**Pull request merged**